### PR TITLE
Add trigger_exception_handler() for post-mortem debugging of caught exceptions

### PR DESCRIPTION
**Pull request by nshepperd**
Adds a convenient trigger_exception_handler() function (previously named post_mortem(); renamed per review discussion) that can be called to immediately trigger the graphical postmortem debugger given any exc_info structure. We need to suspend sys.monitoring callbacks while doing this, because the built in reentrancy protection doesn't apply when pydevd suspends directly from normal code instead of a monitoring callback. Functions were added in _pydevd_sys_monitoring to do that.

### CODE REVIEW

**Comment by nshepperd** [comment]
See https://github.com/microsoft/debugpy/issues/1530.

**Comment by rchiodo** [comment]
Thanks for the submission but I think this would need some sort of automated test before we'd accept it. And maybe a section in the readme.md explaining how to use it.

**Comment by nshepperd** [comment]
Actually I was wondering how to setup tests for this. I figured it out today. Hooray!

**Comment by rchiodo** [comment]
I'm confused about something. pdb.post_mortem attaches the debugger to the process and starts debugging it. This won't. You have to be already attached, right? 

What would be the point of this then? Is it just so you don't have to break on exceptions in the debugger, but instead control when the break happens?

**Comment by rchiodo** [comment]
I think ideally this would actually start VS code and open the debugger. Which might be doable, but would probably have to be configurable somehow.

**Comment by nshepperd** [comment]
Yes, the point is so that you can eg. make a context manager or function decorator to debug exceptions that bubble past it instead of setting bespoke breakpoint, which isn't really directly possible for this. I find something like this indispensable for debugging things like api servers or unit tests (pytest) that have exception handlers that eat exceptions. With this you can just wrap your route handler or set an exception handler with conftest.py and launch the program under the debugger and everything just works.

I think previously debugpy had the 'user uncaught exceptions' breakpoint added for similar use cases, but that's much less specific and in some cases has too many false positives to be useful.

You're right this is a bit different from pdb.post_mortem() in that that actually starts the debugger. I'm not sure about starting vscode though, heh. Maybe it's possible via some sort of vscode:// url? What we could do is at least run debugpy.listen and wait for the debugger to attach? I'm not sure i would actually use that though, usually if i'd want to debug something with vscode I'm already debugging it (eg. I press the 'debug unit test' button) and the issue is just getting the breakpoint.

**Comment by nshepperd** [comment]
Either way each use cases should work, i think. If you wanted to launch vscode and attach the debugger you can use existing debugpy functions for that and then call post_mortem. Or if we include that in the function you can check if debugpy.is_client_connected() before calling it if you don't want that.

**Comment by rchiodo** [comment]
> With this you can just wrap your route handler or set an exception handler with conftest.py and launch the program under the debugger and everything just works.

If this is the intended use case, maybe the name is off then. It's not post_mortem, it's more like 'force_exception_handler' or something like that. I feel like 'post_mortem' is intended to be like pdb means it. It starts a debugger even when it's not already attached.

**Comment by rchiodo** [comment]
Your unit tests also seem to be failing. Not sure if you can see the errors or not

**Comment by nshepperd** [comment]
`trigger_exception_handler()`, maybe?

I see the test failures. Cython issue. I forgot to regenerate the cython files. Confusing because it worked locally.

**Comment by rchiodo** [comment]
`trigger_exception_handler` sounds good to me :)

**Review by rchiodo — APPROVED** [review]

**Comment by rchiodo** [comment]
LGTM. I'll get somebody else to take a look.

**Comment by rchiodo** [comment]
🔒 Automated review in progress — @rchiodo is auto-reviewing this PR.

**Comment by heejaechang** [comment]
🔒 Automated review in progress — @heejaechang is auto-reviewing this PR.

**Review by heejaechang — APPROVED** [review]
Approved via Review Center.

File: src/debugpy/_vendored/pydevd/pydevd.py [hunk 1]
```diff
@@ -2427,6 +2427,51 @@ def do_stop_on_unhandled_exception(self, thread, frame, frames_byid, arg):
...
+        try:
+            if as_uncaught:
+                self.stop_on_unhandled_exception(self, thread, additional_info, excinfo)
+            else:
+                self.do_stop_on_unhandled_exception(thread, user_frame, frames_byid, excinfo)
+        finally:
+            if PYDEVD_USE_SYS_MONITORING:
+                if saved_sys_monitoring_trace:
+                    pydevd_sys_monitoring.resume_current_thread_tracing()
+            additional_info.is_tracing -= 1
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/pydevd.py]: 📍 src/debugpy/_vendored/pydevd/pydevd.py:2455
The Skeptic verified (reproduced with a snippet: `tracing restored? False`) that `suspend_current_thread_tracing()`, `set_additional_thread_info`, and `is_tracing += 1` all run **outside** the `try`. If anything raises in that window — most realistically a `KeyboardInterrupt` during a live debug session, or a failure in `set_additional_thread_info` — the `finally` calling `resume_current_thread_tracing()` never runs, leaving `thread_info.trace = False` permanently and silently killing breakpoints on that thread for the rest of the session. Move the suspend + `is_tracing += 1` setup inside the `try` so the `finally` always restores state.

File: src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 2]
```diff
@@ -1866,6 +1866,56 @@ def stop_monitoring(all_threads=False):
...
+    Use resume_current_thread_tracing() to restore.
+    """
+    try:
+        thread_info = _thread_local_info.thread_info
+    except:
+        # Create the thread info if it doesn't exist yet: if tracing is not
+        # explicitly disabled here, the first monitoring event on this thread
+        # would create it with tracing enabled.
+        thread_info = _get_thread_info(True, 1)
+        if thread_info is None:
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1887]: 📍 src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1875
When the `except` branch freshly creates a `ThreadInfo` via `_get_thread_info(True, 1)` (tracing enabled), it still reports `previous_state = True`. The caller's `finally` then sees `saved == True` and re-enables tracing on a thread that was **not** previously traced, contradicting "restore previous state" and potentially causing unexpected breakpoint stops on background threads. When the `ThreadInfo` is created here rather than found, return `False` so the caller skips the resume. Mirror the same fix in the `.pyx` (and regenerate the `.c`).

File: tests/debugpy/test_trigger_exception_handler.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        exc_info = session.request("exceptionInfo", {"threadId": occ.body['threadId']})
+        assert exc_info == some.dict.containing(
+            {
+                "exceptionId": some.str.matching(r"(.+\.)?ValueError"),
+                "description": "something went wrong",
+                "breakMode": "unhandled",
+            }
+        )
+
+        session.request_continue()
```
Comment (rchiodo) [inline tests/debugpy/test_trigger_exception_handler.py:284]: 📍 tests/debugpy/test_trigger_exception_handler.py:1
The PR's core justification — suspending `sys.monitoring` to avoid reentrancy — is untested. None of the 5 tests set a breakpoint condition that re-enters monitoring during the stop, and none assert that breakpoints still fire on the same thread **after** `request_continue()`. If `suspend_current_thread_tracing()` were a no-op (or the resume-leak bugs triggered), every test would still pass. Add a test that hits a real breakpoint on the same thread after `trigger_exception_handler` returns, proving tracing was correctly restored.

File: src/debugpy/public_api.py [hunk 4]
```diff
@@ -212,6 +212,33 @@ def trace_this_thread(__should_trace: bool):
...
+    Example::
+
+        try:
+            risky_operation()
+        except Exception:
+            debugpy.trigger_exception_handler()  # Uses current exception
+            raise
+    """
+
+
```
Comment (rchiodo) [inline src/debugpy/public_api.py:252]: 📍 src/debugpy/public_api.py:215
The `as_uncaught` parameter is never documented in the public docstring or README, despite changing a non-obvious, inverted-feeling contract: `True` respects the uncaught-exception breakpoint filters (may not stop), while `False` forces post-mortem regardless. Document both `as_uncaught` semantics and note the `as_uncaught=False` + library-only-traceback case.

File: src/debugpy/_vendored/pydevd/pydevd.py [hunk 5]
```diff
@@ -2427,6 +2427,51 @@ def do_stop_on_unhandled_exception(self, thread, frame, frames_byid, arg):
...
+            user_frame = None
+            while tb is not None:
+                frame = tb.tb_frame
+                # Skip debugger-internal frames, use last user frame
+                if self.get_file_type(frame) is None:
+                    user_frame = frame
+                frames.append(frame)
+                tb = tb.tb_next
+
+            if user_frame is None:
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/pydevd.py:2450]: 📍 src/debugpy/_vendored/pydevd/pydevd.py:2430
The `as_uncaught=False` path silently no-ops when the traceback contains only library frames (`user_frame` stays `None` → debug log + return). The README/docstring imply `as_uncaught=False` always enters post-mortem, so a caller wrapping a library-only exception gets a surprising silent non-stop. At minimum document this; consider surfacing it more loudly than a debug-level log.

**Comment by rchiodo** [comment]
GitHub cannot anchor PR review comments to unchanged lines in the diff. Falling back to a general PR comment for [src/debugpy/server/api.py:L396](https://github.com/microsoft/debugpy/blob/92434f46a43292154a6ff10dfff21cf849f4dca2/src/debugpy/server/api.py#L396).

📍 src/debugpy/server/api.py:366
Missing negative-path test coverage for the early returns here: no test covers "no client connected → returns gracefully" or "None arg / no current exception → does nothing". These are cheap to add and lock in the documented no-op semantics.

File: src/debugpy/_vendored/pydevd/pydevd.py [hunk 6]
```diff
@@ -2427,6 +2427,51 @@ def do_stop_on_unhandled_exception(self, thread, frame, frames_byid, arg):
...
+        if PYDEVD_USE_SYS_MONITORING:
+            saved_sys_monitoring_trace = pydevd_sys_monitoring.suspend_current_thread_tracing()
+        thread = threading.current_thread()
+        additional_info = self.set_additional_thread_info(thread)
+        additional_info.is_tracing += 1
+
+        try:
+            if as_uncaught:
+                self.stop_on_unhandled_exception(self, thread, additional_info, excinfo)
+            else:
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/pydevd.py]: 📍 src/debugpy/_vendored/pydevd/pydevd.py:2466
`self.stop_on_unhandled_exception(self, thread, additional_info, excinfo)` is correct — line 765 binds the module-level function as an instance attribute, so the explicit `self` is the intended `py_db` first arg. It is NOT a bug, but it reads like a double-`self` mistake and is a maintenance trap a future reader may "fix" and break. Add a one-line comment (e.g. `# instance attr holds a module-level fn; first arg is py_db`).

File: tests/debugpy/test_trigger_exception_handler.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        occ = session.wait_for_next(
+            Event("stopped") | Event("terminated"),
+        )
+
+        if occ.event == "terminated":
+            pytest.fail("Debuggee exited without hitting breakpoint")
+        
+
+        exc_info = session.request("exceptionInfo", {"threadId": occ.body['threadId']})
+        assert exc_info == some.dict.containing(
```
Comment (rchiodo) [inline tests/debugpy/test_trigger_exception_handler.py:162]: 📍 tests/debugpy/test_trigger_exception_handler.py:152
The stray `print("About to call trigger_exception_handler with excinfo")` in `test_trigger_exception_handler_excinfo` is leftover debugging. Remove it.

**Comment by rchiodo** [comment]
Overall this is a solid, well-tested addition. The main theme worth addressing before/after merge is robustness of the sys.monitoring tracing state restoration: the suspend + is_tracing setup happens outside the try/finally, the 'previous state' value can be misleading for freshly-created ThreadInfo, and the core reentrancy-suspension behavior is not directly exercised by a test. Also a couple of documentation gaps around as_uncaught. None are blocking.

**Review by rchiodo — APPROVED** [review]
Approved via Review Center.

**Comment by rchiodo** [comment]
Okay sorry, I had our automated tools run over it. It found some things to look at if you can. Thanks for the work :)

**Comment by rchiodo** [comment]
You also need to do this before we can accept it:

```
@microsoft-github-policy-service agree [company="{your company}"]
```

File: src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 8]
```diff
@@ -1866,6 +1866,51 @@ def stop_monitoring(all_threads=False):
...
+    """
+    try:
+        thread_info = _thread_local_info.thread_info
+    except:
+        # Create the ThreadInfo if missing; monitoring events create it with
+        # tracing enabled by default, which would defeat the suspension.
+        thread_info = _get_thread_info(True, 1)
+        if thread_info is None:
+            return False
+    previous_state = thread_info.trace
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1889]: When the `except` branch creates a fresh `ThreadInfo` via `_get_thread_info(True, 1)` (tracing enabled by default), the function still falls through to `return previous_state` (= True). The caller's `finally` then re-enables tracing on a thread that was never previously traced, contradicting the documented "returns the previous tracing state" contract. This path is hard to hit in the current flow (explicitly-untraced threads already own a `ThreadInfo` and take the `try` branch), so it's low severity — but the fix is trivial: `return False` from the `except` branch after creating the info. Mirror the change in `_pydevd_sys_monitoring_cython.pyx` and regenerate the `.c`.

File: tests/debugpy/test_trigger_exception_handler.py [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+
+        import debugpy
+
+        def risky_operation():
+            raise ValueError("something went wrong")  # @raise
+
+        try:
+            risky_operation()
+        except ValueError:
+            debugpy.trigger_exception_handler()
```
Comment (rchiodo) [inline tests/debugpy/test_trigger_exception_handler.py:190]: `test_trigger_exception_handler_restores_tracing` is a good guard that tracing is restored after the stop, but it proves *restoration* only — a no-op `suspend_current_thread_tracing()` would still pass it, leaving the PR's central justification (suspending `sys.monitoring` to avoid PEP 669 reentrancy during the stop) unverified. Consider adding a test that re-enters monitoring *during* the stop (e.g. a conditional exception breakpoint whose condition itself hits another breakpoint) so the suppression itself is asserted.

**Comment by rchiodo** [comment]
Looks good overall — approving. Two minor, non-blocking notes below (a trivial self-consistency fix in the new suspend helper, and an optional test to directly exercise the sys.monitoring suppression). Note that several first-pass findings were already addressed in the current diff (the suspend/resume setup is inside the try with saved state initialized to False, and as_uncaught is documented in the public docstring).

**Review by rchiodo — APPROVED** [review]
Approved via Review Center.

File: src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py [hunk 10]
```diff
@@ -159,6 +159,9 @@ def stop_on_unhandled_exception(py_db, thread, additional_info, arg):
     if exception_breakpoint.condition is not None:
         eval_result = py_db.handle_breakpoint_condition(additional_info, exception_breakpoint, user_frame)
         if not eval_result:
+            # The thread may keep running (e.g. via PyDB.trigger_exception_handler),
+            # so don't leave __exception__ in the frame's locals.
+            remove_exception_from_frame(user_frame)
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py]: The added `remove_exception_from_frame(user_frame)` only runs on the `if not eval_result:` path. `add_exception_to_frame` ran unconditionally, so if `handle_breakpoint_condition`/`handle_breakpoint_expression` propagates a `BaseException` (KeyboardInterrupt/SystemExit — ordinary `Exception`s are swallowed by those helpers), `__exception__` stays in the frame's locals. This is newly observable because the thread now survives the stop. Pair the attach/detach in a `try/finally` (or clean up on all early-return paths, including the `expression` branch below this hunk) so `__exception__` never lingers in a live frame.

File: src/debugpy/_vendored/pydevd/pydevd.py [hunk 11]
```diff
@@ -2427,6 +2427,58 @@ def do_stop_on_unhandled_exception(self, thread, frame, frames_byid, arg):
...
+
+            # Innermost frame that isn't debugger-internal.
+            user_frame = None
+            while tb is not None:
+                if self.get_file_type(tb.tb_frame) is None:
+                    user_frame = tb.tb_frame
+                tb = tb.tb_next
+
+            if user_frame is None:
+                pydev_log.debug("trigger_exception_handler: no user frame found in traceback")
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/pydevd.py]: The comment says `do_stop_on_unhandled_exception` doesn't use `frames_byid`, yet a meaningful-looking `frames_byid = {id(user_frame): user_frame}` is still constructed and passed. This misleads future readers. Pass `{}`/`None` while keeping the comment, or drop the comment if the dict is in fact consumed.

**Comment by rchiodo** [comment]
Looks good overall — solid implementation and test coverage. Left a few minor, non-blocking notes; safe to merge once you've glanced at them.

**Review by rchiodo — APPROVED** [review]
Approved via Review Center.

**Review by heejaechang — APPROVED** [review]
Approved via Review Center.

File: src/debugpy/_vendored/pydevd/pydevd.py (hunk 1 above)
Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/pydevd.py]: Rearranged the try/catch to fix this.

File: src/debugpy/public_api.py (hunk 4 above)
Comment (nshepperd) [reply src/debugpy/public_api.py:252]: fixed

File: src/debugpy/_vendored/pydevd/pydevd.py (hunk 11 above)
Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/pydevd.py]: padding None now. it's indeed unused.

File: tests/debugpy/test_trigger_exception_handler.py [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+            pytest.fail("Debuggee exited without hitting breakpoint")
+
+        exc_info = session.request("exceptionInfo", {"threadId": occ.body["threadId"]})
+        assert exc_info == some.dict.containing(
+            {
+                "exceptionId": some.str.matching(r"(.+\.)?ValueError"),
+                "description": "something went wrong",
+                "breakMode": "unhandled",
+            }
+        )
```
Comment (rchiodo) [inline tests/debugpy/test_trigger_exception_handler.py:48]: 📍 tests/debugpy/test_trigger_exception_handler.py:40

The Rules reviewer flagged that four tests (`_basic`, `_basic_with_exception`, `_excinfo`, `_not_as_uncaught`) assert with `some.dict.containing({...})` and `some.str.matching(r"(.+\.)?ValueError")` without justification, violating `docs/pylancewiki/review/tests-no-partial-asserts.md` (partial-dict/substring matches require a comment explaining why the result cannot be made deterministic). This is a confirmed rule violation and is non-downgradeable. These `some.*` matchers are debugpy's idiomatic harness, and `exceptionId` legitimately varies across Python versions (`ValueError` vs `builtins.ValueError`) — so the minimal compliant fix is to add a short comment stating why exact equality isn't possible (and tighten to `==` on the parts of the `exceptionInfo` response that are deterministic).

[verified]

File: src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 13]
```diff
@@ -1866,6 +1866,51 @@ def stop_monitoring(all_threads=False):
...
+    Suspends tracing for the current thread and returns the previous state,
+    to be restored with resume_current_thread_tracing().
+    """
+    try:
+        thread_info = _thread_local_info.thread_info
+    except:
+        # Create the ThreadInfo if missing; monitoring events create it with
+        # tracing enabled by default, which would defeat the suspension.
+        thread_info = _get_thread_info(True, 1)
+        if thread_info is None:
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1887]: 📍 src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1875

Unresolved from the prior round (kept as warning). When the `except` branch creates a fresh `ThreadInfo` via `_get_thread_info(True, 1)` (trace defaults to `True`), the function still returns `previous_state = True`, contradicting the docstring's "returns the previous state" for a thread that had no prior state. As the Advocate/Architect traced, the naive `return False` is NOT a clean fix (it strands the thread at `trace=False` or defeats suppression), so reconcile the **docstring/contract** instead — e.g. document that a freshly-created info reports `True` by design and `False` only when the thread cannot be resolved. Mirror any wording change in `_pydevd_sys_monitoring_cython.pyx`.

[verified]

File: src/debugpy/server/api.py [hunk 14]
```diff
@@ -364,3 +365,47 @@ def trace_this_thread(should_trace):
...
+        )
+
+    exctype, value, tb = excinfo
+    if exctype is None or value is None or tb is None:
+        log.debug("trigger_exception_handler() ignored - no exception info")
+        return
+
+    if not (
+        isinstance(exctype, type)
+        and isinstance(value, BaseException)
```
Comment (rchiodo) [inline src/debugpy/server/api.py:396]: 📍 src/debugpy/server/api.py:380

Unresolved from the prior round. The identical multi-line `"excinfo must be an exception instance or a (type, value, traceback) tuple ..."` message is raised by both the tuple-shape guard and the element-type guard. When the second guard fires the tuple shape is already correct, so the message is misleading. Hoist it into a single constant, or give the element-type guard a distinct message (e.g. reporting the actual element types).

[verified]

File: tests/debugpy/test_trigger_exception_handler.py [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+            expected_frames=[some.dap.frame(code_to_debug, line="bp")],
+        )
+        session.request_continue()
+
+
+def test_trigger_exception_handler_no_op_paths(pyfile, target, run):
+    """Bad input raises ValueError; calling with no current exception is a no-op."""
+
+    @pyfile
+    def code_to_debug():
```
Comment (rchiodo) [inline tests/debugpy/test_trigger_exception_handler.py:213]: 📍 tests/debugpy/test_trigger_exception_handler.py:190

Unresolved coverage gap. `test_trigger_exception_handler_restores_tracing` proves tracing is *restored* after the stop, but a no-op `suspend_current_thread_tracing()` would still pass it — leaving the PR's central justification (suppressing PEP 669 reentrancy *during* the stop) unasserted. Consider adding a test that re-enters monitoring while the stop is in progress (e.g. a conditional exception breakpoint whose condition itself hits another breakpoint) so the suppression itself is validated.

[verified]

File: src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py [hunk 16]
```diff
@@ -1866,6 +1866,51 @@ def stop_monitoring(all_threads=False):
...
+            return False
+    previous_state = thread_info.trace
+    thread_info.trace = False
+    return previous_state
+
+
+# fmt: off
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1894]: 📍 src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1887
The `except` branch still creates `thread_info` via `_get_thread_info(True, 1)` (default `trace=True`) and returns `previous_state = True`, contradicting the docstring's "returns the previous state" for a thread that was never traced (Skeptic, Architect, and Rules all re-confirm this). **However**, the Advocate's caller-trace shows the literal `return False` requested last round is not a clean fix: returning `False` *before* setting `thread_info.trace = False` skips suppression during the stop (defeating the PR's PEP-669 purpose on this path), while returning `False` *after* suspending would leave the fresh thread permanently at `trace=False` (a suppression leak). The current behavior — suppress during the stop, then resume to the thread's `True` default — is actually functionally correct. Resolve the contract honestly: either clarify the docstring that for a freshly-created `ThreadInfo` "previous state" is the thread's default (enabled), or, if a fresh thread should end untraced, suspend during the stop AND return `False` so resume is skipped. Mirror whatever you choose in the `.pyx` and regenerate the `.c`.

[verified]

File: src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py (hunk 2 above)
Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/_pydevd_sys_monitoring/_pydevd_sys_monitoring.py:1887]: I think how it is currently is actually correct. True is the default state (for everything except pydevd-daemon threads), in the sense that every other monitoring event sets thread_info like that if it doesn't exist. So that's what we should restore.

File: src/debugpy/_vendored/pydevd/pydevd.py (hunk 5 above)
Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/pydevd.py:2450]: I think this is fine. The meaning of "library frames" is very narrow here (a static list of stdlib files) and this basically would never happen. I'll make it a warn() log though.

File: tests/debugpy/test_trigger_exception_handler.py (hunk 15 above)
Comment (nshepperd) [reply tests/debugpy/test_trigger_exception_handler.py:213]: Most of the tests already fail if suspend/resume is commented out due to LINE events occurring inside the stop corrupting debugpy state (which is why i originally implemented all this), so there's no need for a specific reentrancy test really.

File: src/debugpy/server/api.py (hunk 14 above)
Comment (nshepperd) [reply src/debugpy/server/api.py:396]: seems to be a pylance thing

**Comment by nshepperd** [comment]
I think this is good now. Thank Claude for letting me stop procrastinating on this. I'll get CLA permission from my employer.

**Comment by rchiodo** [comment]
Looks good overall. A couple of low-priority, non-blocking notes on a docstring/contract mismatch and a duplicated error message; safe to merge.

**Review by rchiodo — APPROVED** [review]
Approved via Review Center.

**Review by heejaechang — APPROVED** [review]
Approved via Review Center.

File: src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py [hunk 17]
```diff
@@ -155,14 +155,19 @@ def stop_on_unhandled_exception(py_db, thread, additional_info, arg):
...
+    try:
+        if exception_breakpoint.condition is not None:
+            eval_result = py_db.handle_breakpoint_condition(additional_info, exception_breakpoint, user_frame)
+            if not eval_result:
+                return
+
+        if exception_breakpoint.expression is not None:
+            py_db.handle_breakpoint_expression(exception_breakpoint, additional_info, user_frame)
+    finally:
+        remove_exception_from_frame(user_frame)
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py]: On Python 3.14, an existing `__exception__` local makes `remove_exception_from_frame()` raise `ValueError`, aborting the default stop and potentially masking the caught exception after an unconditional stop. Preserve and restore any previous binding rather than unconditionally removing it, with regression tests for both `as_uncaught` modes.

File: src/debugpy/_vendored/pydevd/pydevd.py [hunk 18]
```diff
@@ -2427,6 +2427,56 @@ def do_stop_on_unhandled_exception(self, thread, frame, frames_byid, arg):
...
+            # Innermost frame that isn't debugger-internal.
+            user_frame = None
+            while tb is not None:
+                if self.get_file_type(tb.tb_frame) is None:
+                    user_frame = tb.tb_frame
+                tb = tb.tb_next
+
+            if user_frame is None:
+                pydev_log.warn("trigger_exception_handler: no user frame found in traceback")
+                return
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/pydevd.py:2452]: Using `get_file_type(...) is None` excludes `LIB_FILE` frames as well as debugger-internal frames; a library-only traceback therefore never stops despite `as_uncaught=False`. Exclude only `PYDEV_FILE` frames so unconditional mode can select ordinary library frames.

File: src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py (hunk 17 above)
Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py]: Oh, you're right. It does fail on 3.14. I've fixed this with a context manager that restores the previous state, using it in both places.

File: src/debugpy/_vendored/pydevd/pydevd.py (hunk 18 above)
Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/pydevd.py:2452]: Sure.

File: src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py (hunk 17 above)
Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py]: Interestingly, Claude says this is also a problem on HEAD. It shows me this repro testcase for the userUnhandled exception handler, which seemingly hangs the unit test unless the context manager/restore method is used there:

```py
import pytest

from tests import debug
from tests.patterns import some
from tests.timeline import Event


def test_uu_repro(pyfile, target, run):
    @pyfile
    def code_to_debug():
        import threading

        import debuggee

        debuggee.setup()

        def thread_target():
            __exception__ = "sentinel"  # real local in the user boundary frame
            raise ValueError("boom")  # @raise

        t = threading.Thread(target=thread_target)
        t.start()
        t.join()
        print("main continued after thread")  # @done

    with debug.Session() as session:
        session.config["justMyCode"] = True
        session.expected_exit_code = some.int
        with run(session, target(code_to_debug)):
            session.request(
                "setExceptionBreakpoints", {"filters": ["userUnhandled"]}
            )
            session.set_breakpoints(code_to_debug, ["done"])

        session.wait_for_stop("exception")
        session.request_continue()
        session.wait_for_next(Event("exited") | Event("terminated"))
```

Comment (rchiodo) [reply src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py]: The tests do run in 3.14 as far as I know, so it should be working in main

Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_breakpoints.py]: Right, but you didn't have a unit test that puts an existing local named `__exception__`, as far as i can see.

**Comment by nshepperd** [comment]
@microsoft-github-policy-service agree company="Anlatan"

File: src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py [hunk 19]
```diff
@@ -43,6 +43,32 @@ def remove_exception_from_frame(frame):
...
+
+
+class exception_on_frame(object):
+    """Exposes '__exception__' in the frame's locals during a stop, then restores
+    whatever was there before. Restoring by assignment (rather than deleting) avoids
+    a ValueError on 3.14+, where FrameLocalsProxy forbids deleting a real local -- as
+    happens when the user's own code has a variable named '__exception__'."""
+
+    def __init__(self, frame, exception_info):
+        self.frame = frame
```
Comment (rchiodo) [inline src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py:56]: Hmm, I'm wondering if we can just stick the exception on the frame as an attribute. Modifying the locals is usually not so great because then the user sees it in the locals list (and the potential conflict). Although it did that before. But that might explain the error you were talking about with 3.14.

Something like this instead:

```python
self.frame.__stored_exception = exception_info
```

Then removing it would just be deleting that attribute.

Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py:56]: TBH I'm not sure what `__exception__` is used for. I assumed the point was for users to see it in the locals list (eg. vscode debugging terminal). Or maybe it's a legacy thing used by Eclipse?

Comment (rchiodo) [reply src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py:56]: Looking at it, it seems it takes special precautions to remove it from the locals just for that reason. So that users don't see it. I think we can probably just change addExceptionToFrame and removeExceptionFromFrame by just adding an attribute on it. Then we wouldn't have to special case locals.

Comment (nshepperd) [reply src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py:56]: That can't be the case because in actual fact I do see it when i use the debugger (with vanilla debugpy release). For example, in vscode's python debugger, `__exception__` appears as "special variable" under the locals in the Variables pane. And afaict it's not used for anything internally, so if we were going to make it a frame attribute (no longer visible externally) we might as well delete it entirely. The purpose of the cleanup after the stop is so when the thread continues running (eg. userUnhandled exception path where it's not fatal) it doesn't clutter your locals during unrelated later stops.

File: src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py [hunk 20]
```diff
@@ -43,6 +43,32 @@ def remove_exception_from_frame(frame):
...
+    def __enter__(self):
+        self._previous = self.frame.f_locals.get("__exception__", _NO_EXCEPTION)
+        add_exception_to_frame(self.frame, self.exception_info)
+        return self
+
+    def __exit__(self, *args):
+        if self._previous is _NO_EXCEPTION:
+            remove_exception_from_frame(self.frame)
+        else:
+            self.frame.f_locals["__exception__"] = self._previous
```
Comment (heejaechang) [inline src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py:68]: If a stopped user frame declares `__exception__` but has not assigned it before raising, `f_locals.get()` returns `_NO_EXCEPTION`. `add_exception_to_frame()` then binds that real fast local, and `__exit__` takes the delete path; deleting a real local through Python 3.14's `FrameLocalsProxy` raises `ValueError`. Handle declared-but-unbound locals without deleting them, and add coverage for that case.

**Review by rchiodo — APPROVED** [review]

File: src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py (hunk 19 above)
Comment (rchiodo) [reply src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py:56]: Oh yeah it does seem it's actually on purpose.

Comment (rchiodo) [reply src/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_frame_utils.py:56]: Like that's its only purpose. To expose exception when stopping

**Comment by rchiodo** [comment]
Thanks for the PR. I'm going to merge.
